File: sdk/typescript/src/result.ts
import { TableData } from './table-data';

/**
 * Test execution status
 */
export enum TestStatus {
  PASSED = 'PASSED',
  FAILED = 'FAILED',
  SKIPPED = 'SKIPPED',
  ERROR = 'ERROR',
  UNSUPPORTED = 'UNSUPPORTED',
}

/**
 * Result of a single test case execution
 */
export class ComplianceResult {
  constructor(
    public readonly testId: string,
    public readonly status: TestStatus,
    public outputData?: TableData,
    public errorMessage?: string,
    public errorDetails?: string,
    public executionTimeMs: number = 0
  ) {}

  /**
   * Create a result with output data
   */
  withOutput(data: TableData): ComplianceResult {
    return new ComplianceResult(
      this.testId,
      this.status,
      data,
      this.errorMessage,
      this.errorDetails,
      this.executionTimeMs
    );
  }

  /**
   * Create a result with error message
   */
  withError(message: string): ComplianceResult {
    return new ComplianceResult(
      this.testId,
      this.status,
      this.outputData,
      message,
      this.errorDetails,
      this.executionTimeMs
    );
  }

  /**
   * Create a result with error details
   */
  withErrorDetails(details: string): ComplianceResult {
    return new ComplianceResult(
      this.testId,
      this.status,
      this.outputData,
      this.errorMessage,
      details,
      this.executionTimeMs
    );
  }

  /**
   * Create a result with execution time
   */
  withExecutionTime(ms: number): ComplianceResult {
    return new ComplianceResult(
      this.testId,
      this.status,
      this.outputData,
      this.errorMessage,
      this.errorDetails,
      ms
    );
  }

  /**
   * Check if test passed
   */
  isPassed(): boolean {
    return this.status === TestStatus.PASSED;
  }

  /**
   * Check if test failed
   */
  isFailed(): boolean {
    return this.status === TestStatus.FAILED;
  }

  /**
   * Check if test was skipped
   */
  isSkipped(): boolean {
    return this.status === TestStatus.SKIPPED;
  }

  /**
   * Check if test had an error
   */
  isError(): boolean {
    return this.status === TestStatus.ERROR;
  }

  /**
   * Check if test is unsupported
   */
  isUnsupported(): boolean {
    return this.status === TestStatus.UNSUPPORTED;
  }
}

/**
 * Aggregated report for a test suite execution
 */
export class ComplianceReport {
  private results: ComplianceResult[] = [];

  /**
   * Add a test result to the report
   */
  addResult(result: ComplianceResult): void {
    this.results.push(result);
  }

  /**
   * Get all results
   */
  getResults(): ComplianceResult[] {
    return this.results;
  }

  /**
   * Get total number of tests
   */
  totalCount(): number {
    return this.results.length;
  }

  /**
   * Get number of passed tests
   */
  passedCount(): number {
    return this.countByStatus(TestStatus.PASSED);
  }

  /**
   * Get number of failed tests
   */
  failedCount(): number {
    return this.countByStatus(TestStatus.FAILED);
  }

  /**
   * Get number of skipped tests
   */
  skippedCount(): number {
    return this.countByStatus(TestStatus.SKIPPED);
  }

  /**
   * Get number of error tests
   */
  errorCount(): number {
    return this.countByStatus(TestStatus.ERROR);
  }

  /**
   * Get number of unsupported tests
   */
  unsupportedCount(): number {
    return this.countByStatus(TestStatus.UNSUPPORTED);
  }

  /**
   * Calculate pass rate (0-100)
   */
  passRate(): number {
    if (this.results.length === 0) return 0;
    return (this.passedCount() / this.results.length) * 100;
  }

  /**
   * Get total execution time in milliseconds
   */
  totalExecutionTimeMs(): number {
    return this.results.reduce((sum, r) => sum + r.executionTimeMs, 0);
  }

  /**
   * Check if all tests passed
   */
  allPassed(): boolean {
    return this.passedCount() === this.totalCount();
  }

  /**
   * Get summary string
   */
  summary(): string {
    return `Passed: ${this.passedCount()}/${this.totalCount()} (${this.passRate().toFixed(1)}%)`;
  }

  private countByStatus(status: TestStatus): number {
    return this.results.filter((r) => r.status === status).length;
  }
}

// Made with Bob

File: sdk/go/result.go
package compliance

import (
	"fmt"
	"time"
)

// TestStatus represents the execution status of a test
type TestStatus string

const (
	TestStatusPassed      TestStatus = "PASSED"
	TestStatusFailed      TestStatus = "FAILED"
	TestStatusSkipped     TestStatus = "SKIPPED"
	TestStatusError       TestStatus = "ERROR"
	TestStatusUnsupported TestStatus = "UNSUPPORTED"
)

// ComplianceResult represents the result of a single test case execution
type ComplianceResult struct {
	TestID         string
	Status         TestStatus
	OutputData     *TableData
	ErrorMessage   string
	ErrorDetails   string
	ExecutionTime  time.Duration
}

// NewComplianceResult creates a new result with the given test ID and status
func NewComplianceResult(testID string, status TestStatus) *ComplianceResult {
	return &ComplianceResult{
		TestID: testID,
		Status: status,
	}
}

// WithOutput sets the output data
func (r *ComplianceResult) WithOutput(data *TableData) *ComplianceResult {
	r.OutputData = data
	return r
}

// WithError sets the error message
func (r *ComplianceResult) WithError(message string) *ComplianceResult {
	r.ErrorMessage = message
	return r
}

// WithErrorDetails sets detailed error information
func (r *ComplianceResult) WithErrorDetails(details string) *ComplianceResult {
	r.ErrorDetails = details
	return r
}

// WithExecutionTime sets the execution time
func (r *ComplianceResult) WithExecutionTime(duration time.Duration) *ComplianceResult {
	r.ExecutionTime = duration
	return r
}

// IsPassed checks if the test passed
func (r *ComplianceResult) IsPassed() bool {
	return r.Status == TestStatusPassed
}

// IsFailed checks if the test failed
func (r *ComplianceResult) IsFailed() bool {
	return r.Status == TestStatusFailed
}

// IsSkipped checks if the test was skipped
func (r *ComplianceResult) IsSkipped() bool {
	return r.Status == TestStatusSkipped
}

// IsError checks if the test had an error
func (r *ComplianceResult) IsError() bool {
	return r.Status == TestStatusError
}

// IsUnsupported checks if the test is unsupported
func (r *ComplianceResult) IsUnsupported() bool {
	return r.Status == TestStatusUnsupported
}

// ComplianceReport aggregates results for a test suite execution
type ComplianceReport struct {
	Results []*ComplianceResult
}

// NewComplianceReport creates a new empty report
func NewComplianceReport() *ComplianceReport {
	return &ComplianceReport{
		Results: make([]*ComplianceResult, 0),
	}
}

// AddResult adds a test result to the report
func (r *ComplianceReport) AddResult(result *ComplianceResult) {
	r.Results = append(r.Results, result)
}

// TotalCount returns the total number of tests
func (r *ComplianceReport) TotalCount() int {
	return len(r.Results)
}

// PassedCount returns the number of passed tests
func (r *ComplianceReport) PassedCount() int {
	return r.countByStatus(TestStatusPassed)
}

// FailedCount returns the number of failed tests
func (r *ComplianceReport) FailedCount() int {
	return r.countByStatus(TestStatusFailed)
}

// SkippedCount returns the number of skipped tests
func (r *ComplianceReport) SkippedCount() int {
	return r.countByStatus(TestStatusSkipped)
}

// ErrorCount returns the number of tests with errors
func (r *ComplianceReport) ErrorCount() int {
	return r.countByStatus(TestStatusError)
}

// UnsupportedCount returns the number of unsupported tests
func (r *ComplianceReport) UnsupportedCount() int {
	return r.countByStatus(TestStatusUnsupported)
}

// PassRate calculates the pass rate as a percentage (0-100)
func (r *ComplianceReport) PassRate() float64 {
	if len(r.Results) == 0 {
		return 0.0
	}
	return (float64(r.PassedCount()) / float64(len(r.Results))) * 100.0
}

// TotalExecutionTime returns the sum of all test execution times
func (r *ComplianceReport) TotalExecutionTime() time.Duration {
	var total time.Duration
	for _, result := range r.Results {
		total += result.ExecutionTime
	}
	return total
}

// AllPassed checks if all tests passed
func (r *ComplianceReport) AllPassed() bool {
	return r.PassedCount() == r.TotalCount()
}

// Summary returns a summary string
func (r *ComplianceReport) Summary() string {
	return fmt.Sprintf("Passed: %d/%d (%.1f%%)",
		r.PassedCount(), r.TotalCount(), r.PassRate())
}

func (r *ComplianceReport) countByStatus(status TestStatus) int {
	count := 0
	for _, result := range r.Results {
		if result.Status == status {
			count++
		}
	}
	return count
}

